compiler/dyno/lib/resolution/resolution-queries.cpp

```diff
@@ -1441,6 +1453,56 @@ const TypedFnSignature* instantiateSignature(Context* context,
 
       formalIdx++;
     }
+
+    // At this point, we have computed the instantiated type for this
+    // formal. However, what we're still missing some information,
+    // and furthermore, we have not enforced type query constraints.
+
+    if (entry.isVarArgEntry()) {
+      // Vararg entries don't get substitutions at this point, so
+      // manually update type queries.
+      if (auto vld = formal->toVarLikeDecl()) {
+        if (auto te = vld->typeExpression()) {
+            visitor.resolveTypeQueries(te, useType.type());
+        }
+      }
+    } else {
+      // Substitutions have been updated; re-run resolution to get better
+      // intents, vararg info, and to extract type query info.
+      formal->traverse(visitor);
+      formalType = getProperFormalType(r, entry, ad, formal);
+    }
+
+    // Type queries have now been computed. We need to verify that type
+    // query constraints are matched. To do this, instruct resolver to avoid
+    // using substitutions, and to preserve previously computed type query
+    // info. This way, we'll get as output the type expression's QualifiedType
+    // which incorporates type query info.
+    if (auto vld = formal->toVarLikeDecl()) {
+      if (vld->typeExpression()) {
+        visitor.ignoreSubstitutionFor = formal;
+        visitor.skipTypeQueries = true;
+      }
+    }
+    formal->traverse(visitor);
+    auto qFormalType = getProperFormalType(r, entry, ad, formal);
+
+    if (entry.isVarArgEntry()) {
+      // We only need to canPass the tuple element types.
+      qFormalType = getVarArgTupleElemType(qFormalType);
+    } else {
+      // Explicitly override the type in the resolver to what we have found it
+      // to be before the type-query-aware traversal.
+      r.byAst(entry.formal()).setType(formalType);
+    }
+
+    auto checkType = !useType.isUnknown() ? useType : formalType;
+    // With the type and query-aware type known, make sure that they're compatible
+    auto passResult = canPass(context, checkType, qFormalType);
```

Comment:
shouldn't we be checking the actual's type?

Reply:
`useType` is computed from `canPass` and `actualType`; in the cases where that's somehow unknown (it happens), I fall back to `formalType`. I think this is morally fine because in the end, we want to check if the type we computed for the formal meets the type query requirements, which `canPass` will ensure.